examples/bert/bert_model.py

```diff
@@ -0,0 +1,709 @@
+            self._inner_activation, dtype=policy
+        )
+        self._inner_dropout_layer = keras.layers.Dropout(
+            rate=self._inner_dropout
+        )
+        self._output_dense = keras.layers.experimental.EinsumDense(
+            einsum_equation,
+            output_shape=(None, hidden_size),
+            bias_axes="d",
+            name="output",
+            kernel_initializer=self._kernel_initializer,
+            **common_kwargs
+        )
+        self._hidden_dropout = keras.layers.Dropout(rate=self._hidden_dropout)
+        # Use float32 in layernorm for numeric stability.
+        self._output_layer_norm = keras.layers.LayerNormalization(
+            name="output_layer_norm",
+            axis=-1,
+            epsilon=self._norm_epsilon,
+            dtype=tf.float32,
+        )
+
+        super().build(input_shape)
+
+    def get_config(self):
+        config = {
+            "num_attention_heads": self._num_heads,
+            "inner_size": self._inner_size,
+            "inner_activation": self._inner_activation,
+            "hidden_dropout": self._hidden_dropout_rate,
+            "attention_dropout": self._attention_dropout_rate,
+            "output_range": self._output_range,
+            "kernel_initializer": keras.initializers.serialize(
+                self._kernel_initializer
+            ),
+            "bias_initializer": keras.initializers.serialize(
+                self._bias_initializer
+            ),
+            "kernel_regularizer": keras.regularizers.serialize(
+                self._kernel_regularizer
+            ),
+            "bias_regularizer": keras.regularizers.serialize(
+                self._bias_regularizer
+            ),
+            "activity_regularizer": keras.regularizers.serialize(
+                self._activity_regularizer
+            ),
+            "kernel_constraint": keras.constraints.serialize(
+                self._kernel_constraint
+            ),
+            "bias_constraint": keras.constraints.serialize(
+                self._bias_constraint
+            ),
+            "use_bias": self._use_bias,
+            "norm_first": self._norm_first,
+            "norm_epsilon": self._norm_epsilon,
+            "inner_dropout": self._inner_dropout,
+            "attention_initializer": keras.initializers.serialize(
+                self._attention_initializer
+            ),
+            "attention_axes": self._attention_axes,
+        }
+        base_config = super().get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+
+    def call(self, inputs):
+        """Transformer self-attention encoder block call.
+
+        Args:
+            inputs: a single tensor or a list of tensors.
+                `input tensor` as the single sequence of embeddings.
+                [`input tensor`, `attention mask`] to have the additional
+                attention mask.
+                [`query tensor`, `key value tensor`, `attention mask`] to have
```

Comment:
Rather than nested arguments, we should prefer explicit keyword arguments.

We should also make the layer compatible with the built-in masking system. This means that if the query or key have masked values, we should take them into account in the attention mask. Let's discuss this

Reply:
Yeah, added named arguments, which helps. Using built in masking I left as a TODO, we can clean that up as we make this an actual library layer. For now it takes an `attention_mask` kwarg, which is equivalent to `MultiHeadAttention` in core keras.